src/lib.rs

```diff
@@ -98,43 +120,28 @@ where
         Err(Error::Ack)
     }
 
-    fn start(&mut self) {
-        self.clk.set_low();
-        self.dio.set_high();
-        self.clk.set_high();
+    fn start(&mut self) -> Result<(), Error> {
+        self.clk.set_low().map_err(|_| Error::IO).unwrap();
+        self.dio.set_high().map_err(|_| Error::IO).unwrap();
+        self.clk.set_high().map_err(|_| Error::IO).unwrap();
         self.delay();
-        self.dio.set_low();
+        self.dio.set_low().map_err(|_| Error::IO).unwrap();
+
+        Ok(())
     }
 
-    fn stop(&mut self) {
-        self.clk.set_low();
-        self.dio.set_low();
-        self.clk.set_high();
+    fn stop(&mut self) -> Result<(), Error> {
+        self.clk.set_low().map_err(|_| Error::IO).unwrap();
+        self.dio.set_low().map_err(|_| Error::IO).unwrap();
+        self.clk.set_high().map_err(|_| Error::IO).unwrap();
         self.delay();
-        self.dio.set_high();
+        self.dio.set_high().map_err(|_| Error::IO).unwrap();
         self.delay();
+
+        Ok(())
     }
 
     fn delay(&mut self) {
         self.delay.delay_us(DELAY_USECS);
     }
 }
-
-const MAX_FREQ_KHZ: u16 = 250;
```

Comment:
I don't see the reason for moving these private constants on top of the module. I believe the module should start with interface details, not implementation ones.

Reply:
I find it easier to read with the constant values defined at the beginning before they're used. I just hadn't changed it back but I can revert it, I don't feel strongly about it either way.

Reply:
I'd like to revert it.

Reply:
Great. Now I see the constant got changes from 250 to 500.

But.. that's OK. I managed to wire STM32 in order to check the actual behavior. Works well :+1: